File: forms/medical-language-speaking-assessment-for-english/back-end-with-loco/src/engine/oet_grader.rs
//! OET Speaking Sub-test (Medicine) grader.
//!
//! Pure function: takes an `AssessmentData` value, returns the linguistic
//! total (0-24), clinical total (0-15), raw total (0-39), scaled score
//! (0-500), grade (A, B, C+, C, D, E), the per-criterion audit trail, and
//! the list of fired rules and additional flags.
//!
//! Ported verbatim from `front-end-form-with-html/js/oet-grader.js` and
//! `flagged-issues.js`. Criterion IDs, flag IDs, and grade-band cut-points
//! are preserved.

use super::flagged_issues::detect_additional_flags;
use super::rules::{CriterionDomain, criterion_registry};
use super::types::{AssessmentData, CriterionScore, FiredRule, GradingResult};
use super::utils::{classify_scaled_score, mean_or_single, round1};

/// Pure scoring engine.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let mut per_criterion_scores: Vec<CriterionScore> = Vec::new();
    let mut fired_rules: Vec<FiredRule> = Vec::new();

    let mut linguistic_total: f64 = 0.0;
    let mut clinical_total: f64 = 0.0;

    for c in criterion_registry() {
        match c.domain {
            CriterionDomain::Linguistic => {
                let rp1 = (c.linguistic_accessor)(&data.linguistic_role_play1);
                let rp2 = (c.linguistic_accessor)(&data.linguistic_role_play2);
                let mean = mean_or_single(rp1, rp2);
                per_criterion_scores.push(CriterionScore {
                    id: c.id.to_string(),
                    domain: c.domain.as_str().to_string(),
                    label: c.label.to_string(),
                    max_score: c.max_score,
                    role_play1_score: rp1,
                    role_play2_score: rp2,
                    mean_score: mean.map(round1),
                });
                if let Some(m) = mean {
                    linguistic_total += m;
                    fired_rules.push(FiredRule {
                        id: c.id.to_string(),
                        category: "Linguistic".to_string(),
                        description: c.label.to_string(),
                        score: round1(m),
                    });
                }
            }
            CriterionDomain::Clinical => {
                let v = (c.clinical_accessor)(&data.clinical_indicators);
                per_criterion_scores.push(CriterionScore {
                    id: c.id.to_string(),
                    domain: c.domain.as_str().to_string(),
                    label: c.label.to_string(),
                    max_score: c.max_score,
                    role_play1_score: None,
                    role_play2_score: None,
                    mean_score: v,
                });
                if let Some(score) = v {
                    clinical_total += score;
                    fired_rules.push(FiredRule {
                        id: c.id.to_string(),
                        category: "Clinical Communication".to_string(),
                        description: c.label.to_string(),
                        score,
                    });
                }
            }
        }
    }

    let raw_total = linguistic_total + clinical_total;
    let scaled_score = ((raw_total / 39.0) * 500.0).round() as u32;
    let grade = classify_scaled_score(scaled_score);

    let timestamp = chrono::Utc::now().to_rfc3339();

    // First build a partial result so the flag detector can read the grade
    // / scaled-score / per-criterion table.
    let mut result = GradingResult {
        linguistic_total: round1(linguistic_total),
        clinical_total: round1(clinical_total),
        raw_total: round1(raw_total),
        scaled_score,
        grade,
        per_criterion_scores,
        fired_rules,
        additional_flags: Vec::new(),
        timestamp,
    };

    result.additional_flags = detect_additional_flags(data, &result);

    result
}

File: forms/medical-language-speaking-assessment-for-cymraeg/front-end-with-svelte/src/lib/engine/oet-grader.ts
// Clinical Welsh-language (Cymraeg) speaking assessment grader.
//
// Pure function: takes an `AssessmentData` object and returns the linguistic
// total (0-24), clinical total (0-15), raw total (0-39), scaled score
// (0-500), grade (A, B, C+, C, D, E), the per-criterion audit trail, and the
// list of fired rules. The flagged issues are attached separately by
// `detectAdditionalFlags`.
//
// Calculation:
//   linguisticTotal = sum over the 4 linguistic criteria of
//                     mean(rolePlay1Score, rolePlay2Score)        -> 0..24
//   clinicalTotal   = sum over the 5 clinical indicators           -> 0..15
//   rawTotal        = linguisticTotal + clinicalTotal              -> 0..39
//   scaledScore     = round(rawTotal / 39 * 500)                   -> 0..500
//
// Grade bands (CEFR-mapped cut-points, parallel to the OET Medicine bands):
//   450-500 -> A (CEFR C2), 350-440 -> B (CEFR C1), 300-340 -> C+ (CEFR B2+),
//   200-290 -> C (CEFR B2), 100-190 -> D (CEFR B1), 0-90 -> E (CEFR A2-).

import { criterionRegistry } from './rules';
import { detectAdditionalFlags } from './flagged-issues';
import type {
	AssessmentData,
	CriterionScore,
	FiredRule,
	GradingResult,
	LinguisticField,
	OETGrade
} from './types';

/** Map a 0-500 scaled score to a grade. */
export function classifyScaledScore(score: number): OETGrade {
	if (score >= 450) return 'A';
	if (score >= 350) return 'B';
	if (score >= 300) return 'C+';
	if (score >= 200) return 'C';
	if (score >= 100) return 'D';
	return 'E';
}

/** Coerce a value to a finite number, otherwise return null. */
function toNumOrNull(v: unknown): number | null {
	if (v === null || v === undefined || v === '') return null;
	const n = Number(v);
	return Number.isFinite(n) ? n : null;
}

/**
 * Mean of two role-play scores. Returns null only when *both* are null. If only
 * one is provided it is treated as the criterion mean (single-role-play partial
 * completion still produces a meaningful figure).
 */
function meanOrSingle(a: unknown, b: unknown): number | null {
	const x = toNumOrNull(a);
	const y = toNumOrNull(b);
	if (x === null && y === null) return null;
	if (x === null) return y;
	if (y === null) return x;
	return (x + y) / 2;
}

/** Round to 1 decimal place for display. */
function round1(n: number): number {
	return Math.round(n * 10) / 10;
}

/** Pure scoring engine (without flags). */
export function gradeOET(data: AssessmentData): Omit<GradingResult, 'additionalFlags' | 'timestamp'> {
	const perCriterionScores: CriterionScore[] = [];
	const firedRules: FiredRule[] = [];

	let linguisticTotal = 0;
	let clinicalTotal = 0;

	for (const c of criterionRegistry) {
		if (c.domain === 'linguistic') {
			const field = c.dataField as LinguisticField;
			const rp1 = toNumOrNull(data.linguisticRolePlay1?.[field]);
			const rp2 = toNumOrNull(data.linguisticRolePlay2?.[field]);
			const mean = meanOrSingle(rp1, rp2);
			perCriterionScores.push({
				id: c.id,
				domain: 'linguistic',
				label: c.label,
				maxScore: c.maxScore,
				rolePlay1Score: rp1,
				rolePlay2Score: rp2,
				meanScore: mean === null ? null : round1(mean)
			});
			if (mean !== null) {
				linguisticTotal += mean;
				firedRules.push({ id: c.id, category: 'Linguistic', description: c.label, score: round1(mean) });
			}
		} else {
			const v = toNumOrNull(
				(data.clinicalIndicators as unknown as Record<string, unknown>)?.[c.dataField]
			);
			perCriterionScores.push({
				id: c.id,
				domain: 'clinical',
				label: c.label,
				maxScore: c.maxScore,
				rolePlay1Score: null,
				rolePlay2Score: null,
				meanScore: v
			});
			if (v !== null) {
				clinicalTotal += v;
				firedRules.push({
					id: c.id,
					category: 'Clinical Communication',
					description: c.label,
					score: v
				});
			}
		}
	}

	const rawTotal = linguisticTotal + clinicalTotal;
	const scaledScore = Math.round((rawTotal / 39) * 500);
	const grade = classifyScaledScore(scaledScore);

	return {
		linguisticTotal: round1(linguisticTotal),
		clinicalTotal: round1(clinicalTotal),
		rawTotal: round1(rawTotal),
		scaledScore,
		grade,
		perCriterionScores,
		firedRules
	};
}

/** Full grading result including flags and a timestamp. */
export function gradeAssessment(data: AssessmentData): GradingResult {
	const core = gradeOET(data);
	return {
		...core,
		additionalFlags: detectAdditionalFlags(data, core),
		timestamp: new Date().toISOString()
	};
}
